Mathlib/Init/Data/List/Lemmas.lean

```diff
@@ -0,0 +1,116 @@
+/-
+Copyright (c) 2016 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Leonardo de Moura
+-/
+import Mathlib.Mathport.Rename
+import Std.Data.List.Basic
+import Std.Data.List.Lemmas
+import Mathlib.Init.Data.Nat.Basic
+import Mathlib.Init.Data.Nat.Lemmas
+import Mathlib.Init.Data.List.Basic
+/-!
+Lemmas for `List` not (yet) in `Std`
+-/
+
+
+open List Nat
+
+namespace List
+
+/-- length of list obtained by `map₂` on a pair of lists is the length of the shorter of the two -/
+@[simp]
+theorem length_map₂ (f : α → β → γ) (l₁) : ∀ l₂, length (map₂ f l₁ l₂) =
+    min (length l₁) (length l₂) := by
+  induction l₁ <;> intro l₂ <;> cases l₂ <;>
+    simp [*, add_one, min_succ_succ, Nat.zero_min, Nat.min_zero]
+#align list.length_map₂ List.length_map₂
+
+/-- length of the list consisting of an element repeated `n` times is `n` -/
+@[simp]
+theorem length_repeat (a : α) (n : ℕ) : length («repeat»  a n) = n := by induction n <;> simp [*]
+#align list.length_repeat List.length_repeat
+
+/-- length of the list on removing the `i`th element
+when `i` is less than the length of the original list -/
+theorem length_remove_nth : ∀ (l : List α) (i : ℕ),
+    i < length l → length (removeNth l i) = length l - 1
+  | [], _, _ => rfl
+  | _ :: xs, 0, _ => by
+     simp [removeNth]
+  | x :: xs, i + 1, h => by
+    have : i < length xs := lt_of_succ_lt_succ h
+    dsimp [removeNth]
+    rw [length_remove_nth xs i]
+    show length xs - 1 + 1 = length xs + 1 - 1
+    rw [Nat.add_sub_cancel (length xs) 1]
+    have bd : 1 ≤  length xs := by
+      have bd' : i + 1 ≤ length xs := by assumption
+      have bd'' : 1 ≤ i + 1 := by
+        apply Nat.succ_le_succ
+        apply Nat.zero_le
+      exact Nat.le_trans bd'' bd'
+    exact Nat.sub_add_cancel bd
+    assumption
```

Comment:
Why is this proof much longer than the corresponding lean 3 proof? If proofs break then my instinct is not to fix them, it's to try and figure out why they broke.

Reply:
The difference is length is mainly my lack of skill with tactic proofs. Probably `simp` and `rw` invoked with the correct lemmas will finish after the `dsimp`